File: app/src/main/java/me/ketal/hook/SortAtPanel.java
/* QNotified - An Xposed module for QQ/TIM
 * Copyright (C) 2019-2021 xenonhydride@gmail.com
 * https://github.com/ferredoxin/QNotified
 *
 * This software is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software.  If not, see
 * <https://www.gnu.org/licenses/>.
 */
package me.ketal.hook;

import android.text.TextUtils;

import java.util.Arrays;
import java.util.List;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;
import me.singleneuron.qn_kernel.data.HostInformationProviderKt;
import nil.nadph.qnotified.hook.CommonDelayableHook;
import nil.nadph.qnotified.step.DexDeobfStep;
import nil.nadph.qnotified.util.DexKit;

import static me.ketal.util.TIMVersion.TIM_3_1_1;
import static me.singleneuron.util.QQVersion.QQ_8_1_3;
import static nil.nadph.qnotified.util.Initiator._SessionInfo;
import static nil.nadph.qnotified.util.Initiator.load;
import static nil.nadph.qnotified.util.ReflexUtil.iget_object_or_null;
import static nil.nadph.qnotified.util.ReflexUtil.invoke_virtual;
import static nil.nadph.qnotified.util.Utils.*;

public class SortAtPanel extends CommonDelayableHook {
    public static final SortAtPanel INSTANCE = new SortAtPanel();
    boolean isSort = false;

    protected SortAtPanel() {
        super("ketal_At_Panel_Hook", new DexDeobfStep(DexKit.N_AtPanel__refreshUI), new DexDeobfStep(DexKit.N_AtPanel__showDialogAtView));
    }

    @Override
    protected boolean initOnce() {
        try {
            XposedBridge.hookMethod(DexKit.doFindMethod(DexKit.N_AtPanel__showDialogAtView), new XC_MethodHook() {
                @Override
                protected void afterHookedMethod(MethodHookParam param) {
                    if (!isEnabled())
                        return;
                    String key = (String) param.args[1];
                    isSort = TextUtils.isEmpty(key);
                }
            });
            XposedBridge.hookMethod(DexKit.doFindMethod(DexKit.N_AtPanel__refreshUI), new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    if (!isEnabled())
                        return;
                    Object result = param.args[0];
                    if (isSort) {
                        Object sessionInfo = getFirstByType(param.thisObject, _SessionInfo());
                        String troopUin = iget_object_or_null(sessionInfo, "troopUin", String.class);
                        if (troopUin == null)
                            troopUin = iget_object_or_null(sessionInfo, "a", String.class);
                        Class<?> clzTroopInfo = load("com.tencent.mobileqq.data.troop.TroopInfo");
                        if (clzTroopInfo == null)
                            clzTroopInfo = load("com.tencent.mobileqq.data.TroopInfo");
                        Object troopInfo = invoke_virtual(getTroopManager(), "b", troopUin, String.class, clzTroopInfo);
                        String ownerUin = iget_object_or_null(troopInfo, "troopowneruin", String.class);
                        String[] administrator = iget_object_or_null(troopInfo, "Administrator", String.class).split("\\|");
                        List<String> admin = Arrays.asList(administrator);
                        List<Object> list = getFirstByType(result, List.class);
                        String uin = getUin(list.get(0));
                        boolean isAdmin = "0".equals(uin);
                        Object temp;
                        for (int i = 1; i < list.size(); i++) {
                            Object member = list.get(i);
                            uin = getUin(member);
                            if (uin == null) {
                                throw new NullPointerException("uin == null");
                            }
                            if (uin.equals(ownerUin)) {
                                temp = member;
                                list.remove(member);
                                list.add(isAdmin ? 1 : 0, temp);
                            } else if (admin.contains(uin)) {
                                temp = member;
                                list.remove(member);
                                list.add(isAdmin ? 2 : 1, temp);
                            }
                        }
                    }
                }
            });
            return true;
        } catch (Exception e) {
            log(e);
            return false;
        }
    }

    private String getUin(Object member) {
        String uin = iget_object_or_null(member, "uin", String.class);
        if (uin == null) uin = iget_object_or_null(member, "a", String.class);
        try {
            Long.parseLong(uin);
        } catch (Exception e) {
            return null;
        }
        return uin;
    }

    @Override
    public boolean isValid() {
        if (HostInformationProviderKt.getHostInformationProvider().isTim() && HostInformationProviderKt.getHostInformationProvider().getVersionCode() >= TIM_3_1_1)
            return true;
        else return !HostInformationProviderKt.getHostInformationProvider().isTim() && HostInformationProviderKt.getHostInformationProvider().getVersionCode() >= QQ_8_1_3;
    }
}

File: app/src/main/java/me/ketal/hook/SortAtPanel.kt
/*
 * QNotified - An Xposed module for QQ/TIM
 * Copyright (C) 2019-2021 dmca@ioctl.cc
 * https://github.com/ferredoxin/QNotified
 *
 * This software is non-free but opensource software: you can redistribute it
 * and/or modify it under the terms of the GNU Affero General Public License
 * as published by the Free Software Foundation; either
 * version 3 of the License, or any later version and our eula as published
 * by ferredoxin.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * and eula along with this software.  If not, see
 * <https://www.gnu.org/licenses/>
 * <https://github.com/ferredoxin/QNotified/blob/master/LICENSE.md>.
 */
package me.ketal.hook

import android.view.View
import xyz.nextalone.data.TroopInfo
import xyz.nextalone.util.*
import me.ketal.util.BaseUtil.tryVerbosely
import me.singleneuron.qn_kernel.data.requireMinVersion
import me.singleneuron.qn_kernel.tlb.ConfigTable
import nil.nadph.qnotified.base.annotation.FunctionEntry
import nil.nadph.qnotified.hook.CommonDelayableHook
import nil.nadph.qnotified.step.DexDeobfStep
import nil.nadph.qnotified.util.*
import nil.nadph.qnotified.util.ReflexUtil.getFirstByType

@FunctionEntry
object SortAtPanel : CommonDelayableHook(
    "ketal_At_Panel_Hook",
    DexDeobfStep(DexKit.N_AtPanel__refreshUI),
    DexDeobfStep(DexKit.N_AtPanel__showDialogAtView)
) {
    const val sessionInfoTroopUin = "SortAtPanel.sessionInfoTroopUin"
    private var isSort: Boolean? = null
    override fun initOnce() = tryVerbosely(false) {
        val showDialogAtView = DexKit.doFindMethod(DexKit.N_AtPanel__showDialogAtView)
            ?: DexKit.doFindClass(DexKit.N_AtPanel__showDialogAtView)?.method {
                it.parameterTypes.contentEquals(arrayOf(View::class.java, String::class.java, Boolean::class.java))
            }
        showDialogAtView?.hookAfter(this) {
            isSort = (it.args[1] as String?)?.isNotEmpty()
        }
        val refreshUI = DexKit.doFindMethod(DexKit.N_AtPanel__refreshUI)
            ?: DexKit.doFindClass(DexKit.N_AtPanel__refreshUI)?.method {
                it.parameterTypes.contentEquals(arrayOf("com.tencent.mobileqq.troop.quickat.ui.SearchTask\$SearchResult".clazz))
            }
        refreshUI?.hookBefore(this) {
            if (isSort == true) return@hookBefore
            val sessionInfo = getFirstByType(it.thisObject, Initiator._SessionInfo())
            val troopInfo = TroopInfo(getTroopUin(sessionInfo))
            val list = getFirstByType(it.args[0], MutableList::class.java) as MutableList<Any>
            val isAdmin = "0" == getMemberUin(list[0])
            val admin = mutableListOf<Any>()
            for (i in list.indices) {
                val member = list[i]
                when (getMemberUin(member)) {
                    "0" -> continue
                    troopInfo.troopOwnerUin -> admin.add(0, member)
                    in troopInfo.troopAdmin!! -> admin.add(member)
                }
            }
            list.removeAll(admin)
            list.addAll(if (isAdmin) 1 else 0, admin)
        }
        true
    }

    private fun getTroopUin(sessionInfo: Any?): String? =
        sessionInfo.get("troopUin", String::class.java)
            ?: sessionInfo.get(ConfigTable.getConfig(sessionInfoTroopUin), String::class.java)

    private fun getMemberUin(member: Any?): String? =
        member.get("uin", String::class.java)
            ?: member.get("a", String::class.java)


    override fun isValid(): Boolean = requireMinVersion(QQVersion.QQ_8_1_3, TIMVersion.TIM_3_1_1, PlayQQVersion.PlayQQ_8_2_9)
}
